### ISSUE TITLE
Right click closes a new image edit without warning

### ISSUE THREAD

**Issue by GregariousJB** [ShareX/ShareX#6707]
Apparently you've continued to close this issue for years, despite it still being a problem since it was first mentioned when Greenshot was removed in 2018: #3220, #4822, #3290, #4746, and an honorable mention (still open) at #6487. I stopped looking after these. I'm betting there's more. 4746 mentions a fix but only for captures that have already been edited, which was a great change albeit incomplete.

To duplicate:

1. Create any capture with "Open to image editor" selected in "After capture tasks".
2. Without making any changes in the editor, right-click, either thinking there is a context-menu or just on accident.
3. Cry with varying levels of tears depending on how long it took you to get that capture _just right_ (including an hour or more with a scrolling capture). Your work is gone.

![fuck](https://media.giphy.com/media/wvQIqJyNBOCjK/giphy.gif)

**Comment by themegabyte** [comment]
So I think the "editor" got this change, and not the "after snap instant editor".

I currently just disabled the right click as I am way more prone to hitting the right click as opposed to the `escape` key,

![image](https://user-images.githubusercontent.com/9980686/234984761-e772e38f-ebc9-4544-a695-fee44c84cc38.png)

This will help me with my twitchy fingers for now... 😅

1. But I agree, a confirmation dialogue does have a use case in the After Capture Editor too. 
2. Or instead we could have our shortcut key to always take full screen current monitor snaps so it woul d always save before we can edit it.

I will check for scrolling screenshots to see if I can shoot myself in the foot with the current latest release. 

@GregariousJB

**Comment by GregariousJB** [comment]
> So I think the "editor" got this change, and not the "after snap instant editor".
> 
> I currently just disabled the right click as I am way more prone to hitting the right click as opposed to the `escape` key,
> 
> ![image](https://user-images.githubusercontent.com/9980686/234984761-e772e38f-ebc9-4544-a695-fee44c84cc38.png)
> 
> This will help me with my twitchy fingers for now... 😅
> 
>     1. But I agree, a confirmation dialogue does have a use case in the After Capture Editor too.
> 
>     2. Or instead we could have our shortcut key to always take full screen current monitor snaps so it woul d always save before we can edit it.
> 
> 
> I will check for scrolling screenshots to see if I can shoot myself in the foot with the current latest release.
> 
> @GregariousJB

That setting helps, thanks. Insane that it's not default, or like you said, has a confirmation. Good lord.

**Comment by the-Arioch** [comment]
> So I think the "editor" got this change, and not the "after snap instant editor".

Great catch!

I also wish the "after capture" action list was re-orderable, if it turns out those actions are modal (mutually blocking)

For example GreenShot's sequence there is

1. Save the PNG file (GS can not do video captures, at all :-( )
2. Open the quick-actons menu or the editor
3. In the editor you now can close the window - the file is already saved

Every time you save the file - regardless automatically or from menu, with custom name or auto-name - the path to file is copied to clipboard.
I am not even sure if it is customizable or not - it is so handy that i never had urge to check.

Also interesting the terminology, in GS that all is called not "after captue tasks" but "screenshot destination". Not sure if this is mere make up or is somehow changing developers' mindset or programs' structure.

I did not even notice I could "select several tasks" in GS. Yes, technically i could,...
...perhaps i should give it a test. I just never needed it because Quuick Tasks Menu is actually all i needed in practice.

**Comment by GregariousJB** [comment]
After editing an image's text box, I accidentally hit ENTER twice - once to confirm a text edit and close the text-editing window, then again accidentally - and the whole edit just disappeared. The entire window closed. So, I went digging - what other keys will close the window indiscriminately, causing loss of work? 

Hence, the title of this issue has been updated. Right-click isn't the only thing without any confirmation. This is WORSE, because this can happen AFTER you've made changes to the image.

**Comment by themegabyte** [comment]
Confirmed, Enter actuates the current action.

And if you are at the normal arrow tool it just closes everything which is weird since the default action for the "arrow" tool is supposed to be "do nothing"

**Comment by YakovL** [comment]
I'd like to second this. There should be at least an option(s) to set a confirmation dialog for these cases, better with the default value(s) of true. I've lost a valueable piece of information because of this issue when I tried to reopen a note to edit with enter (closed the whole thing instead).

I suggest to put this proposition (show a confirmation dialog by default/at least add an option for it) to the issue description itself, so that it's clear what is proposed. Note: I haven't even found if it's possible to prevent Enter from firing "Continue".

**Comment by YakovL** [comment]
@Jaex  thanks for fixing one case in 16.1.0: now pressing Esc in annotation (text) editor doesn't at least close the whole window, loosing the captured image and annotations. I hope that more cases will be covered in the future, for instance with proper defaults (all those "do nothing" actions like for right click should be something like "do nothing on confirm" with a proper dialog before throwing the results of the work away).

**Comment by mydoghasfleas** [comment]
Hi, I also just want to ask if this could please be made a feature, that Enter does not automatically close the image editor.

My workflow is simply: after capture, I immediately open the captured image in the image editor.

It would be nice if the "Enter" key could be configured not to close the window, as this is my main gripe; I have to be careful not to accidentally press it and lose what I was painstakingly working on.

Thank you!

**Comment by indigofairyx** [comment]
haha, wtf, when googling this its seemingly been a reported issue easily going back to at least 2019. i expect after a minimum of 6 years they're not going to change it. 

furthermore i changed this setting as mentioned in the post above, thank you @themegabyte 

saved\exported my settings from prefs

did a windows reinstall a month later

reloaded my saved settings.....

and this apparently isn't saved in the settings export either. hahah 🤣

**Comment by GregariousJB** [comment]
I retested the keys that close the editor without warning. Looks like right-click and ESC keys were fixed to give you a warning, but spacebar, ENTER, and (oddly) the backtick/tilda key will still close the editor immediately. There may be other keys - I didn't test all of them.

**Comment by YakovL** [comment]
@GregariousJB for me, ESC closes the window without saving (v16.1). Do you have some setting customized for this behavior?

**Comment by GregariousJB** [comment]
> @GregariousJB for me, ESC closes the window without saving (v16.1). Do you have some setting customized for this behavior?

Are you making any change to your screenshot in the editor first?

An argument could be made that having the screenshot close even when not making changes is frustrating, but this issue is about closing the editor after making changes without any warning, which ignores any work done to it.

**Comment by YakovL** [comment]
Right, makes sense! Indeed it asks if there are some edits.

I guess, it'd be valueable to have a "table" of issues (e.g. actions like doing right click) and workarounds/statuses in the opening post, like:

* ESC – ✅ ok (asks for confirmation before exit, if any edits are done);
* SPACE – ⚠️ bad (closes without saving, even if some edits are done);
* ...

Last updated: 07.2024 (v16.1)

**Comment by the-Arioch** [comment]
That's an opportunity.

Fork it, fox it, and sell for premium :-)

**Comment by indigofairyx** [comment]
I might use autohotkey to block those keys from getting to the window

**Comment by indigofairyx** [comment]
that was easy, here ya go, put this in your AutoHotkey script.

the following buttons will no longer close you're the editor window.
when pressed they will send a mouse click rather than that key.

```ahk
#ifwinactive ShareX - Image editor 
space::click
~::click
rbutton::click
enter::click
#ifwinactive
```

**Comment by mydoghasfleas** [comment]
The way I have solved the issue for myself, for now, is to adjust my workflow as follows:

First, by enabling "Multi Region Mode" in the capture screen:

![image](https://github.com/user-attachments/assets/d0300509-2f1a-4f70-a9bc-2a0dc78b3ce8)

(This is in any case a useful mode for doing more complex captures).

Then I do all my annotations and editing directly in the capture screen.

In my after-capture tasks, I have opted to show the quick menu.

![image](https://github.com/user-attachments/assets/286305f1-c76a-4017-b73a-d0dd16d5376b)

This way, after the capturing and annotation/editing, when I press Enter, I still get the quick menu, from where I just copy the result to my clipboard.

![image](https://github.com/user-attachments/assets/a8370783-a01a-442c-b0d2-744c4764b056)

It's really just a workaround for the Enter-closing-the-window problem, but for now it works for me.

**Comment by GregariousJB** [comment]
Just happened again. Was working on an image, adding notes, arrows, and circles, and accidentally hit ENTER twice on an annotation and the entire thing just disappeared.

Also noticed the ShareX main window is taunting me with it, showing me what it just removed without allowing me to reopen it to continue where I left off:

![fuckthis](https://i.imgur.com/Ed6RtFO.png)

I used to recommend this program to others.

**Comment by GregariousJB** [comment]
This is utterly fucking stupid that it's necessary, but it seems to be working: ![ihatethis](https://i.imgur.com/BhWIb2z.png)

**Comment by YakovL** [comment]
I have to admit, the last comment is even more confusing after the edit. Could you explain what does that setting do and if it helps to deal with the issue? I mean, how can it disable enter outside ShareX, it's like you can't type enter (linebreak) in an ordinary editor?

**Comment by GregariousJB** [comment]
> I have to admit, the last comment is even more confusing after the edit. Could you explain what does that setting do and if it helps to deal with the issue? I mean, how can it disable enter outside ShareX, it's like you can't type enter (linebreak) in an ordinary editor?

I thought my ENTER key was sticking because it wasn't working in several programs. That list of hotkeys has total control over the keyboard, apparently, for as long as ShareX is running. ENTER key **_ded_** when set to "none" as shown.

**Comment by YakovL** [comment]
🤯🙏

**Comment by GregariousJB** [comment]
Don't think it's a bug, actually. ShareX needs access to certain key combos at all times in order to take several types of screenshots. It was just unexpected at first.

**Comment by peterryanhouse** [comment]
> that was easy, here ya go, put this in your AutoHotkey script.
> 
> the following buttons will no longer close the editor window. when pressed they will send a mouse click rather than that key.
> 
> #ifwinactive ShareX - Image editor 
> space::click
> ~::click
> rbutton::click
> enter::click
> #ifwinactive

Where is the "AutoHotkey script" located and how do I add this?

**Comment by peterryanhouse** [comment]
I was able to fix the "Enter" issue but not the Spacebar.  In the "Image Editor", the settings on the top right has an option  to "Auto Copy to Clipboard".  This way, if you press Enter by mistake, it still copies my markups to the clipboard.  

I also disabled the right click as stated above by themegabyte on Apr 27, 2023

**Comment by peterryanhouse** [comment]
> that was easy, here ya go, put this in your AutoHotkey script.
> 
> the following buttons will no longer close the editor window. when pressed they will send a mouse click rather than that key.
> 
> #ifwinactive ShareX - Image editor 
> space::click
> ~::click
> rbutton::click
> enter::click
> #ifwinactive

This disables the buttons if you want to use it when editing text in the screenshot... right??

**Comment by peterryanhouse** [comment]
I found a work around that fixes the spacebar losing your work issue. I am not 100% it works, but it does. The way I figured it out was actually to solve another issue I was having with ShareX. The issue was that it was not saving my original screenshot AND my edited screenshot. I found a reddit string that someone used ShareX as an external editor. When I implemented this, the space bar no longer exited the editor causing lost work. You actually have to click the "X" button to get out of the editor. Also, by using the ShareX editor as an "external editor", it saves a copy of my screenshot every time I hit the spacebar or enter buttons, which is quite nice.

It took a minute to setup, but am very happy I did.

Here is a shared link to my ShareX settings: https://brookssolutions-my.sharepoint.com/:u:/g/personal/phouse_brooksair_com/ET6sRAdo9kZMu_XKb0-pJugBXuPcEXEufsjb3DqEWSzg_Q?e=zPypgV

Here is the reddit thread:
https://www.reddit.com/r/sharex/comments/we0s4m/any_way_to_save_screenshot_then_open_the_image/

Here is the full reddit thread text:

Edited 3y ago
I actualy found a way. Because i always want my screenshots to be saved, then the editor open (even if 20% of the time, i'll press 'esc').i you use the "execute after task", the Saving of the file does not occur when you select the "open in editor".

This is why we need a workaround which is the following:

Use ShareX as an "external editor" using the command line call that opens with editor mode.

ShareX.exe -ImageEditor "filepath".

-ImageEditor "$input"

Now your file will be saved + you have the editor mode :)

see example: https://i.imgur.com/XKNeifo.png

u/amit510 avatar
amit510

1y ago
Thank you for this, finally got it to work. Something to note that there needs to be space added in front of the dash:
-ImageEditor "$input"

https://imgur.com/a/7eqIjtS

u/peterryanhouse1 avatar
peterryanhouse1

37m ago
This worked for me too! Thanks!

Comment Image

u/peterryanhouse1 avatar
peterryanhouse1

36m ago
I used ChatGPT to get my comparable executable file path link with the "%" symbols.

**Comment by YakovL** [comment]
> Where is the "AutoHotkey script" located and how do I add this?

You can put .ahk scripts or links to them into the Startup folder (that's the usual usage), but basically once AHK is installed, you can run them from anywhere, by double-clicking them

**Comment by YakovL** [comment]
This is unfortunate :(

**Comment by GregariousJB** [comment]
@Jaex , before it was closed (prematurely IMHO), this issue was tied at 13th place for most-thumbed-up issue out of ~570 open issues. Perhaps you'd consider reopening it?
